Add a GenericConstantStdVectorMaterial to use std::vector<Real> properties and related material property output

GiudGiud opened the pull request:
RealVectorValue is handled by GenericConstantRealVectorValue, which isnt following our naming convention.

If this is accepted I can add the stateful material property storage support later. The actions have the same issue of not knowing the size of the properties

GiudGiud committed:
Convert the GenericConstantVectorMaterial to use std::vector<Real> properties
RealVectorValue is handled by GenericConstantRealVectorValue

GiudGiud committed:
Add a basic test for the std::vector<Real> material properties

GiudGiud committed:
Add 1st component output for std::vector<Real> properties
refs #31519

GiudGiud committed:
Add support for outputting all components of a vector Real matprop
refs #31519

GiudGiud committed:
Better and safer material output behavior

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3214246/), step Docs: sync website on cdc6d65 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31518/site)

This comment will be updated on new commits.

GiudGiud committed:
Avoid hard-deprecating the GenericConstantVectorMaterial
Use the StdVector particulate instead

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3214423/), step Generate coverage on 661edda wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9676bb</th>
      <th colspan=4>#31518 661edd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.71%</td>
      <td>85.71%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>118183</td>
      <td>118236</td>
      <td>+53</td>
      <td>64</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19698</td>
      <td>19705</td>
      <td>+7</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31518/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31518/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31518/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/src/materials/GenericConstantStdVectorMaterial.C:
```diff
@@ -0,0 +1,87 @@
+    _prop_names(getParam<std::vector<std::string>>("prop_names")),
+    _prop_values(getParam<std::vector<std::vector<Real>>>("prop_values"))
+{
+  unsigned int num_names = _prop_names.size();
+  unsigned int num_values = _prop_values.size();
```
```suggestion
  const auto num_names = _prop_names.size();
  const auto num_values = _prop_values.size();
```

lindsayad commented on framework/src/materials/GenericConstantStdVectorMaterial.C:
```diff
@@ -0,0 +1,87 @@
+  _num_props = num_names;
+  _properties.resize(num_names);
+
+  for (unsigned int i = 0; i < _num_props; i++)
```
```suggestion
  for (const auto i : make_range(_num_props))
```

lindsayad commented on framework/src/materials/GenericConstantStdVectorMaterial.C:
```diff
@@ -0,0 +1,87 @@
+{
+  for (unsigned int i = 0; i < _num_props; i++)
+  {
+    (*_properties[i])[_qp].resize(_prop_values[i].size());
+    for (const auto j : index_range(_prop_values[i]))
+      (*_properties[i])[_qp][j] = _prop_values[i][j];
```
```suggestion
    auto & prop_out = (*_properties[i])[_qp];
    const auto & prop_in = _prop_values[i];
    prop_out.resize(prop_in.size());
    for (const auto j : index_range(prop_in))
      prop_out[j] = prop_in[j];
```

lindsayad commented on test/tests/materials/generic_materials/tests:
```diff
@@ -78,4 +78,20 @@
+      type = 'Exodiff'
+      input = 'generic_constant_std_vector_real.i'
+      exodiff = 'generic_constant_std_vector_real_out.e'
+      detail = 'with constant values.'
```
```suggestion
      detail = 'with constant values, and'
```

GiudGiud committed:
Address Alex review

- more const
- more auto
- keep const ref of property names
- size_t for size ints

Also:
- add link to std::vector matprop mat in both realvectorvalue matprop mat
- correct templating for matoutputaction

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3224669/), step Clang format on 851b704 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31518/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31518/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9e48167076b690bb1d56eea388137d51bb434451`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3224669/), step Clang format on 851b704 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31518/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31518/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9e48167076b690bb1d56eea388137d51bb434451`

GiudGiud commented:
Test failures are unrelated.
Thanks for the review!
